include/souper/Infer/AbstractInterpreter.h

```diff
@@ -16,29 +16,75 @@
-    llvm::KnownBits sle(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
+    llvm::KnownBits add(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits addnsw(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits sub(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits subnsw(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits mul(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits udiv(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits urem(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits and_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits or_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits xor_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits shl(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits lshr(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ashr(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits eq(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ne(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ult(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits slt(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ule(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits sle(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
   }
+
+  bool isConcrete(souper::Inst *I,
+                  bool ConsiderConsts = true,
+                  bool ConsiderHoles = true);
+
+  class KnownBitsAnalysis {
+    std::unordered_map<Inst*, llvm::KnownBits> KBCache;
+
+    // Checks the cache or instruction metadata for knonwbits information
+    bool cacheHasValue(Inst *I);
+
+  public:
+    llvm::KnownBits findKnownBits(Inst *I,
+                                  ConcreteInterpreter &CI,
+                                  bool PartialEval = true);
+
+    static llvm::KnownBits findKnownBitsUsingSolver(Inst *I,
+                                                    Solver *S,
+                                                    std::vector<InstMapping> &PCs);
+
+    static std::string knownBitsString(llvm::KnownBits KB);
+
+    static llvm::KnownBits getMostPreciseKnownBits(llvm::KnownBits A, llvm::KnownBits B);
+  };
+
+  class ConstantRangeAnalysis {
+    std::unordered_map<Inst*, llvm::ConstantRange> CRCache;
+
+    // checks the cache or instruction metadata for cr information
+    bool cacheHasValue(Inst *I);
+
+  public:
+    llvm::ConstantRange findConstantRange(souper::Inst *I,
+                                          ConcreteInterpreter &CI,
+                                          bool PartialEval = true);
```

Comment:
same here-- if the only use case for turning off partial eval is testing, then we should just remove the parameter and adapt the test cases, unless there's a compelling argument why testing is better with partial evaluation turned off

Reply:
@manasij7479

Reply:
Feel free to adapt the tests, not very useful to keep this feature.

Reply:
can you take a stab at this @pranavk ?

Reply:
I just remembered that we also use this flag from the souper-interpret binary. The souper-interpret binary interprets first with partialEval off and second with on. If you think it's a useless feature, I will go ahead removing this PartialEval flag completely.

Reply:
yes, I think this flag clutters up way too much code for the benefit it provides